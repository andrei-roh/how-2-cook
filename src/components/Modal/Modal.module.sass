@use 'src/style/colors'
@import 'src/style/typography'

.modalWrapper
    width: 100%
    height: 100vh
    z-index: 100
    background-color: colors.$fog
    transition: all .3s ease-out
    box-sizing: border-box
    position: fixed
    top: 0
    left: 0

    .modalContainer
        z-index: 200
        position: absolute
        top: 50%
        left: 50%
        transform: translate(-50%, -50%)
        background-color: colors.$white
        width: 80%
        max-width: 320px
        padding: 24px
        border-radius: 12px
        display: flex
        flex-direction: column
        gap: 18px

        .modalMessage
            @extend .T1
            text-align: center

        .buttonsPanel
            width: 100%
            margin-top: 12px
            display: flex
            flex-direction: row
            justify-content: space-between
            gap: 18px

            .modalButton
                @extend .H6
                width: 120px
                height: 42px
                box-sizing: border-box

            .modalCancelButton
                color: colors.$purple
                background-color: colors.$white
                border: 1px solid colors.$purple

                &:hover
                    color: colors.$white
                    background-color: colors.$blue
                    border: none

            .modalSubmitButton
                background-color: colors.$blue
                color: colors.$white

                &:hover
                    background-color: colors.$sky