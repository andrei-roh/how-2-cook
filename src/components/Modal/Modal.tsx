import css from './Modal.module.sass';
import { getClassesList } from 'src/utils/getClassesList';
import CircularProgress from '@mui/material/CircularProgress';
import Button from '@mui/material/Button';

interface ModalProps {
  cancelButtonMessage: string;
  submitButtonMessage: string;
  isLoading: boolean;
  handleClose: () => void;
  handleSubmit: () => void;
  message: string;
  submitClassName?: string;
  cancelClassName?: string;
}

export const Modal = ({
  cancelButtonMessage,
  submitButtonMessage,
  isLoading,
  handleClose,
  handleSubmit,
  message,
  submitClassName,
  cancelClassName,
}: ModalProps) => {
  const submitClassNames = getClassesList(`${css.modalButton} ${css.modalSubmitButton}`, submitClassName);
  const cancelClassNames = getClassesList(`${css.modalButton} ${css.modalCancelButton}`, cancelClassName);

  return (
    <div className={css.modalWrapper}>
      <div className={css.modalContainer}>
        <div className={css.modalMessage}>{message}</div>
        <div className={css.buttonsPanel}>
          <Button onClick={handleClose} className={cancelClassNames}>
            {cancelButtonMessage}
          </Button>
          <Button
            onClick={handleSubmit}
            className={submitClassNames}
            disabled={isLoading}
          >
            {isLoading ? <CircularProgress color='inherit' size={32} /> : submitButtonMessage}
          </Button>
        </div>
      </div>
    </div>
  );
};
