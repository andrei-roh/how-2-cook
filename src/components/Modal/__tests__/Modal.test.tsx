import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { Modal } from '../Modal';

const cancelButtonMessageMock = 'Test Cancel';
const submitButtonMessageMock = 'Test Submit';
const handleCloseMock = vi.fn();
const handleSubmitMock = vi.fn();
const messageMock = 'Test Message';
const submitClassNameMock = 'submitTestClass';
const cancelClassNameMock = 'cancelTestClass';

describe('Testing: Modal', () => {
  it('should render', () => {
    const { container } = render(
      <Modal
        cancelButtonMessage={cancelButtonMessageMock}
        submitButtonMessage={submitButtonMessageMock}
        isLoading={false}
        handleClose={handleCloseMock}
        handleSubmit={handleSubmitMock}
        message={messageMock}
        submitClassName={submitClassNameMock}
        cancelClassName={cancelClassNameMock}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should call provided function after click submit', () => {
    render(
      <Modal
      cancelButtonMessage={cancelButtonMessageMock}
      submitButtonMessage={submitButtonMessageMock}
      isLoading={false}
      handleClose={handleCloseMock}
      handleSubmit={handleSubmitMock}
      message={messageMock}
      submitClassName={submitClassNameMock}
      cancelClassName={cancelClassNameMock}
    />
    );

    const renderedButton = screen.getByText(submitButtonMessageMock);

    fireEvent.click(renderedButton);

    expect(handleSubmitMock).toHaveBeenCalled();
  });

  it('should call provided function after click cancel', () => {
    render(
      <Modal
      cancelButtonMessage={cancelButtonMessageMock}
      submitButtonMessage={submitButtonMessageMock}
      isLoading={false}
      handleClose={handleCloseMock}
      handleSubmit={handleSubmitMock}
      message={messageMock}
      submitClassName={submitClassNameMock}
      cancelClassName={cancelClassNameMock}
    />
    );

    const renderedButton = screen.getByText(cancelButtonMessageMock);

    fireEvent.click(renderedButton);

    expect(handleCloseMock).toHaveBeenCalled();
  });
});
