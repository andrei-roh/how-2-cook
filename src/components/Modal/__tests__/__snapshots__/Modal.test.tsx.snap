// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing: Modal > should render 1`] = `
<div>
  <div
    class="_modalWrapper_c49097"
  >
    <div
      class="_modalContainer_c49097"
    >
      <div
        class="_modalMessage_c49097"
      >
        Test Message
      </div>
      <div
        class="_buttonsPanel_c49097"
      >
        <button
          class="_buttonWrapper_89d844 _modalSubmitButton_c49097 cancelTestClass"
        >
          Test Cancel
        </button>
        <button
          class="_buttonWrapper_89d844 _modalSubmitButton_c49097 submitTestClass"
        >
          Test Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
